### PR TITLE
Ensure that block execution maintains L1 compatibility with optimism feature enabled

### PR DESCRIPTION
**Pull request by BrianBland**
* Only fetch L1 block info and compute l1 cost when the chain spec includes an Optimism config
* Only transfer L1 Cost and Base Fee to the respective vaults when the chain spec includes an Optimism config

Addresses the following tests (#31):
```
reth-revm executor::tests::sanity_execution
reth-revm executor::tests::test_selfdestruct
```

Follow-up: Add new test cases which cover the cases where the chain has Optimism enabled

### CODE REVIEW

File: crates/revm/src/executor.rs [hunk 1]
```diff
@@ -333,24 +345,31 @@ where
...
-                    U256::from(
-                        block
-                            .base_fee_per_gas
-                            .unwrap_or_default()
-                            .saturating_mul(result.gas_used()),
-                    ),
-                    &mut post_state,
-                )?;
+                if self.chain_spec.optimism.is_some() {
+                    match l1_cost {
```
Comment (merklefruit) [inline crates/revm/src/executor.rs:349]: lint: please use `if let Some(l1_cost)` instead of match

File: crates/revm/src/executor.rs [hunk 2]
```diff
@@ -263,20 +268,27 @@ where
...
-                // Check if the sender balance can cover the L1 cost.
-                // Deposits pay for their gas directly on L1 so they are exempt from the L2 tx fee.
-                if !transaction.is_deposit() {
-                    if sender_account.info.balance.cmp(&l1_cost) == std::cmp::Ordering::Less {
-                        return Err(BlockExecutionError::InsufficientFundsForL1Cost {
-                            have: sender_account.info.balance.to::<u64>(),
-                            want: l1_cost.to::<u64>(),
-                        })
+                match l1_cost {
+                    Some(l1_cost) => {
```
Comment (merklefruit) [inline crates/revm/src/executor.rs:272]: lint: please use `if let Some(l1_cost)` instead of match

File: crates/revm/src/executor.rs [hunk 3]
```diff
@@ -250,7 +254,8 @@ where
             #[cfg(feature = "optimism")]
             {
                 let db = self.db();
-                let l1_cost = l1_block_info.calculate_tx_l1_cost(transaction);
+                let l1_cost =
+                    l1_block_info.map(|l1_block| l1_block.calculate_tx_l1_cost(transaction));
```
Comment (merklefruit) [inline crates/revm/src/executor.rs:258]: consider using `if let` here too for readibility:

```rust
let l1_cost = if let Some(l1_block) = l1_block_info {
    Some(l1_block.calculate_tx_l1_cost(transaction))
} else {
    None
};
```

File: crates/revm/src/executor.rs [hunk 4]
```diff
@@ -333,24 +345,31 @@ where
...
-                            .base_fee_per_gas
-                            .unwrap_or_default()
-                            .saturating_mul(result.gas_used()),
-                    ),
-                    &mut post_state,
-                )?;
+                if self.chain_spec.optimism.is_some() {
+                    match l1_cost {
+                        Some(l1_cost) => {
+                            // Route the l1 cost and base fee to the appropriate optimism vaults
```
Comment (merklefruit) [inline crates/revm/src/executor.rs:351]: please move this comment above the `if self.chain_spec.optimism` line, as it explains both increments

File: crates/revm/src/optimism.rs [hunk 5]
```diff
@@ -21,6 +21,7 @@ const NON_ZERO_BYTE_COST: u64 = 16;
 /// uint64 _sequenceNumber, bytes32 _batcherHash, uint256 _l1FeeOverhead, uint256 _l1FeeScalar)
 ///
 /// For now, we only care about the fields necessary for L1 cost calculation.
+#[derive(Clone, Copy)]
```
Comment (merklefruit) [inline crates/revm/src/optimism.rs:24]: why is this needed? where are we cloning it? Not a big deal anyway, but just want to make sure it's needed

**Review by merklefruit — CHANGES_REQUESTED** [review]
minor comments, overall lgtm!

File: crates/revm/src/optimism.rs (hunk 5 above)
Comment (BrianBland) [reply crates/revm/src/optimism.rs:24]: Now that `l1_block_info` is an Optional type, the inner `L1BlockInfo` must be either moved or copied when it's accessed in each loop iteration. If we don't derive clone/copy, then the compiler rejects the use of a moved value.

Comment (merklefruit) [reply crates/revm/src/optimism.rs]: Awesome!

**Review by merklefruit — APPROVED** [review]
LGTM!
